File: emcrypt-core/src/main/java/com/beam/emcryptcore/model/inbox/Email.java
package com.beam.emcryptcore.model.inbox;

import com.beam.emcryptcore.base.Base;
import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;
import org.springframework.data.annotation.Transient;
import org.springframework.data.annotation.TypeAlias;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.Collection;
import java.util.List;
import java.util.stream.Stream;

@Document("Email")
@TypeAlias("Email")
@AllArgsConstructor
@NoArgsConstructor
@Accessors(chain = true)
@Data
@Builder
public class Email extends Base {

    private String identifier; // tenant

    private Subject from;
    private List<Recipient> to;
    private List<Recipient> cc;
    private List<Recipient> bcc;


    private String subject;

    private String key; // encrypted key
    private String data; // encrypted data

    private Options options;

    private List<Attachment> attachments;

    @JsonIgnore
    @Transient
    public List<Recipient> recipients(){
        return Stream.of(to, cc, bcc)
                .flatMap(Collection::stream)
                .toList();
    }
}

File: emcrypt-web/src/main/ui/src/app/modules/user/model/email.ts
import { Base } from "src/app/common/models/model";
import { Attachment, Recipient, Subject } from "../../reader/model/decrypted";
import { Options } from "../../reader/model/options";

export class Email extends Base {
  identifier: string;
  from: Subject;
  to: Array<Recipient>;
  cc: Array<Recipient>;
  bcc: Array<Recipient>;
  all: Array<Recipient>;
  subject: String;
  options: Options;
  attachments: Array<Attachment>;

  init(_email: Email) {
    super.init(_email);
    Object.assign(this, _email);

    this.from = Object.assign(new Subject(), _email.from);

    this.to = new Array<Recipient>();
    _email.to.forEach((rcp) => {
      this.to.push(Object.assign(new Recipient(), rcp));
    });

    this.cc = new Array<Recipient>();
    _email.cc.forEach((rcp) =>
      this.cc.push(Object.assign(new Recipient(), rcp))
    );

    this.bcc = new Array<Recipient>();
    _email.bcc.forEach((rcp) => {
      this.bcc.push(Object.assign(new Recipient(), rcp));
    });
    this.all = this.to.concat(this.cc).concat(this.bcc);

    this.options = Object.assign(new Options(), _email.options);

    this.attachments = new Array<Attachment>();
    _email.attachments.forEach((att) => {
      this.attachments.push(Object.assign(new Attachment(), att));
    });
  }
}
